GitLab authentication backend

BenoitKnecht opened the pull request:
This feature was requested in #5623, but it was closed after `auth.generic_oauth` was introduced in #4718.

However, as @kri5 mentioned in #5623, `auth.generic_oauth` doesn't let admins filter access based on GitLab group membership, which limits its usefulness in the case of a public GitLab instance.

This PR introduces an `auth.gitlab` backend which is very similar to `auth.github` and introduces an `allowed_groups` option that takes a list of allowed groups.

Fixes #5623

BenoitKnecht committed:
social: add GitLab authentication backend

GitLab could already be used as an authentication backend by properly
configuring `auth.generic_oauth`, but then there was no way to authorize
users based on their GitLab group membership.

This commit adds a `auth.gitlab` backend, similar to `auth.github`, with
an `allowed_groups` option that can be set to a list of groups whose
members should be allowed access to Grafana.

BenoitKnecht committed:
docs: document GitLab authentication backend

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/grafana/grafana?pullRequest=11314) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/grafana/grafana?pullRequest=11314) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/grafana/grafana?pullRequest=11314) it.</sub>

BenoitKnecht commented:
@bergquist  I would love to this picked up in 5.2. Please let me know if there's anything I can do to help make it happen.

BenoitKnecht commented:
Just a friendly reminder that it would be great to see GitLab authentication support in 5.3, and that I would be more than happy to work on this PR if someone gets a chance to review it.

marefr commented on pkg/social/gitlab_oauth.go:
```diff
@@ -0,0 +1,131 @@
+		return nil, fmt.Errorf("User %s is inactive", data.Username)
+	}
+
+	userInfo := &BasicUserInfo{
```
Think you'll need to set the id here to support if email address are change in gitlab for example then Grafana would be able to handle that automatically.

BenoitKnecht replied:
Thank you very much for the review! I pushed a new comment to address this.

marefr commented:
Thanks for the reminder. We're sorry it taken some time for us to come back to this. Did take a quick look at the code and it looks good, added a comment.

Currently, the plan is that this will be included in 5.3 if nothing unexpected happens.

BenoitKnecht committed:
social: gitlab_oauth: set user ID in case email changes

Set `BasicUserInfo.Id` in the value returned by
`SocialGitlab.UserInfo()`, in case the email address of the user changes
in GitLab. That way, the user association won't be lost in Grafana.

marefr commented on docs/sources/installation/configuration.md:
````diff
@@ -428,6 +428,102 @@ allowed_organizations = github google
+http://203.0.113.31:3000/login/gitlab
+```
+
+Finally, select *api* as the *Scope* and submit the form. You'll get an
````
Can be good to mention that **read_user** scope can be used instead of **api** when allowed_groups not are used.

BenoitKnecht replied:
Thanks for the suggestion, added.

marefr reviewed (changes requested):
Seems to work as expected :+1: 

Have just a minor comment on the documentation that would be nice to add.

BenoitKnecht committed:
docs: gitlab: add note about more restrictive API scope

If `allowed_groups` is not used with GitLab authentication, the
*read_user* scope can be used instead of *api*.

marefr committed:
docs: add grafana version note for gitlab oauth

marefr reviewed (approved):
Excellent. Great job! Thanks

marefr commented:
Thank you for contributing to Grafana!
